src/net/client/cache.rs

```diff
@@ -0,0 +1,1160 @@
+    C: Clock + Send + Sync,
+{
+    /// State of the request.
+    state: RequestState,
+
+    /// The request message.
+    request_msg: CR,
+
+    /// The upstream transport of the connection.
+    upstream: Upstream,
+
+    /// The cache of the connection.
+    cache: Cache<Key, Arc<Value<C>>>,
+
+    /// The configuration of the connection.
+    config: Config,
+
+    /// The clock to use for expiring cache entries.
+    clock: C,
+}
+
+impl<CR, Upstream, C> Request<CR, Upstream, C>
+where
+    CR: Clone + ComposeRequest + Send + Sync,
+    Upstream: SendRequest<CR> + Send + Sync,
+    C: Clock + Debug + Send + Sync + 'static,
+{
+    /// Create a new Request object.
+    fn new(
+        request_msg: CR,
+        upstream: Upstream,
+        cache: Cache<Key, Arc<Value<C>>>,
+        config: Config,
+        clock: C,
+    ) -> Request<CR, Upstream, C> {
+        Self {
+            state: RequestState::Init,
+            request_msg,
+            upstream,
+            cache,
+            config,
+            clock,
+        }
+    }
+
+    /// This is the implementation of the get_response method.
+    // This function is cancel safe.
+    async fn get_response_impl(&mut self) -> Result<Message<Bytes>, Error> {
+        loop {
+            match &mut self.state {
+                RequestState::Init => {
+                    let msg = self.request_msg.to_message()?;
+                    let header = msg.header();
+                    let opcode = header.opcode();
+
+                    // Extract Qname, Qclass, Qtype
+                    let mut question_section = msg.question();
+                    let question = match question_section.next() {
+                        None => {
+                            // No question. Just forward the request.
+                            let request = self
+                                .upstream
+                                .send_request(self.request_msg.clone());
+                            self.state =
+                                RequestState::GetResponseNoCache(request);
+                            continue;
+                        }
+                        Some(question) => question?,
+                    };
+                    if question_section.next().is_some() {
+                        // More than one question. Just forward the request.
+                        let request = self
+                            .upstream
+                            .send_request(self.request_msg.clone());
+                        self.state =
+                            RequestState::GetResponseNoCache(request);
+                        continue;
+                    }
+                    let qname = question.qname();
+                    let qclass = question.qclass();
+                    let qtype = question.qtype();
+
+                    if !(opcode == Opcode::Query && qclass == Class::In) {
+                        // Anything other than a query on the Internet class
```

Comment:
Is this limitation worth mentioning in the `cache.rs` module level docs?

Reply:
I would find it very useful to have that text in the comment rather than "More than one question. Just forward the request." as it is now, as the current text leaves the reader having to guess why this decision has been made (even if they think they know, they can't know the mind of the developer who wrote the code/comment and so may be wrong).